lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -4282,3 +4326,27 @@ AnyFunctionType *swift::adjustFunctionTypeForConcurrency(
 bool swift::completionContextUsesConcurrencyFeatures(const DeclContext *dc) {
   return contextUsesConcurrencyFeatures(dc);
 }
+
+bool swift::isAsynchronousContext(const DeclContext *dc) {
+  if (auto func = dyn_cast<AbstractFunctionDecl>(dc)) {
+    return func->isAsyncContext();
+  }
+
+  if (auto closure = dyn_cast<AbstractClosureExpr>(dc)) {
+    if (auto type = closure->getType()) {
+      if (auto fnType = type->getAs<AnyFunctionType>()) {
```

Comment:
When we're inside the constraint solver, this will give the wrong answer because the closures we encounter won't have types yet. Rather, we will need to look up the closure's type within the constraint system to see if it's `async`.

Reply:
Good catch! I have overlooked this one, I guess we need a callback to get a type of a closure here...

Reply:
Fixed!